### PR TITLE
prov/rxm: Rework locking around progress

### PR DESCRIPTION
**Pull request by shefty**
RxM can drive progress across an EP from several call sites.  The result is that while serialization is provided on individual queues, structures, and lists, the overall progress model is not thread safe.  This can be seen by the app as receiving messages out of order, or received messages being associated with the wrong posted buffer.

To fix this, we need to serialize the progress handling at a higher level.  However, this requires restructuring the locking, so that the EP lock can be held across the full handling of messages.  Care must be taken that we don't lose locks where they are needed, and that we don't introduce nested locking/deadlocks.

This series provides several patches to restructure and simplify the code.  We then move the EP locking up to the desired functions, so that all progress is serialized on an EP.  Progress can be initiated as part of completion processing, driven by connection processes (EQ events), or started as part of a transfer.

### CODE REVIEW

**Comment by shefty** [comment]
@swelch - This series reworks the locking in the rxm provider.  I *think* it's correct and closes all races.  The EP locking is moved up closer to the level of the API, rather than trying to provide fine grained locking around specific structures/queues/lists.  This should fix races handing MSG completions and associating those with specific posted receive buffers.  Locking around the atomic operations at the EP level is included.

@dmitrygladkov- please review when you return.  Because lower-level locks were removed, this should provide some benefit when locking is not needed (fewer no-op calls).  The drawback is that multiple threads accessing the same rxm EP will be serialized at a higher level.  That seems acceptable, since the EP is a single resource.

I tried to maintain a mix of broad versus fine grained locking, but the code paths are too complex to do this without introducing a lot of uncertainty regarding correctness.

**Comment by swelch** [comment]
Thanks @shefty - I will have time to look it over later this evening/first thing in the morning, and will also pull the PR down and run/test with it.

**Comment by rajachan** [comment]
The OSU MPI benchmark suite has a multi-threaded latency test (osu_latency_mt) where multiple threads within a rank post receives. Might be useful in testing this PR.

**Comment by swelch** [comment]
@shefty - The changes look good/made sense to me. I also pulled down the PR and tested with the OSU benchmarks (including osu_latency_mt with a receive thread count of 10) and RC/XRC and didn't have issues.

**Comment by shefty** [comment]
Thanks @swelch !  I hadn't gotten to running those benchmarks yet.

File: prov/rxm/src/rxm_ep.c [hunk 1]
```diff
@@ -1306,19 +1209,23 @@ rxm_ep_inject_send_fast(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
...
 		inject_pkt->hdr.size = len;
 		memcpy(inject_pkt->data, buf, len);
-		return rxm_ep_msg_inject_send(rxm_ep, rxm_conn, inject_pkt,
+		ret = rxm_ep_msg_inject_send(rxm_ep, rxm_conn, inject_pkt,
 					      pkt_size, rxm_ep->util_ep.tx_cntr_inc);
 	} else {
-		return rxm_ep_emulate_inject(rxm_ep, rxm_conn, buf, len, pkt_size,
-					     inject_pkt->hdr.data, inject_pkt->hdr.flags,
-					     inject_pkt->hdr.tag, inject_pkt->hdr.op);
+		ofi_ep_lock_acquire(&rxm_ep->util_ep);
```
[the remark this answers is not in the document]
Comment (shefty) [reply prov/rxm/src/rxm_ep.c]: Thanks - I'll update and repush

**Pull request merged**